openmc/search.py

```diff
@@ -0,0 +1,249 @@
+from numbers import Real
+from types import FunctionType
+
+import openmc
+import openmc.model
+from openmc.checkvalue import check_type, check_iterable_type, check_length, \
+    check_value
+
+
+_SCALAR_BRACKETED_METHODS = ['brentq', 'brenth', 'ridder', 'bisect']
+
+
+class KeffSearch(object):
+    """Class to perform a keff search by modifying a model parametrized by a
+    single independent variable.
+
+    Parameters
+    ----------
+    model_builder : FunctionType
+        Callable function which builds a model according to a passed
+        parameter. This function must return an openmc.model.Model object.
+    guess : Real, optional
+        Initial guess for the parameter to be searched in
+        :param:`model_builder`. One of :param:`guess` or :param`bracket` must
+        be provided.
+    target_keff : Real, optional
+        keff value to search for, defaults to 1.0.
+    bracket : None or Iterable of Real, optional
+        Bracketing interval to search for the solution; if not provided,
+        a generic non-bracketing method is used. If provided, the brackets
+        are used. Defaults to no brackets provided. One of :param:`guess` or
+        :param`bracket` must be provided. If both are provided, the bracket
+        will be preferentially used.
+
+    Attributes
+    ----------
+    model_builder : FunctionType
+        Callable function which builds a model according to parameters passed.
+        This function must return an openmc.model.Model
+        object.
+    initial_guess : Iterable of Real
+        Initial guess for the parameter to be searched in
+        :param:`model_builder`.
+    target_keff : Real
+        keff value to search for.
+    bracket : None or Iterable of Real
+        Bracketing interval to search for the solution; if not provided,
+        a generic non-bracketing method is used. If provided, the brackets
+        are used.
+    model_args : dict
+        Keyword-based arguments to pass to the :param:`model_builder` method.
+    print_iterations : bool
+        Whether or not to print the guess and the resultant keff during the
+        iteration process. Defaults to False.
+    print_output : bool
+        Whether or not to print the OpenMC output during the iterations.
+        Defaults to False.
+    guesses : List of Real
+        List of guesses attempted by the search
+    keffs : List of Real
+        List of keffs corresponding to the guess attempted by the search
+    keff_uncs : List of Real
+        List of keff uncertainties corresponding to the guess attempted by the
+        search
+
+    """
+
+    def __init__(self, model_builder, guess=None, bracket=None,
+                 target_keff=1.0, model_args={}):
```

Comment:
We actually do not need a whole class for this now and a function would definitely suffice.  The reason why this is a class is because at first I was planning on making a generic Search class and then have extended types for KeffSearch, ResponseSearch (a generic response function), or maybe even a TallySearch (or whatever else one would want with the generic base class).  

The developer extending the class would then just need to replace the `_search function` with whatever is needed to get the job done.

Clearly I did not do that here.  I wanted to first get far enough to see how this would rack out with the most likely use-case (99% of all searches using the neutron transport equation are k-eigenvalue searches in reality as far as I'm aware).

Let me know if this class-based route seems useful; if so it makes the most sense to leave this as a class now.  Also let me know if you'd prefer the generic class be implemented now, I can do that as well.